scio-google-cloud-platform/src/main/scala/com/spotify/scio/bigquery/client/BigQuery.scala

```diff
@@ -217,7 +219,42 @@ object BigQuery {
 
     def credentials: Credentials = _credentials
 
-    lazy val underlying: Bigquery = {
+    def execute[T](fn: Bigquery => AbstractGoogleClientRequest[T]): T = {
+      def getAuthenticatedUser: (String, String) = {
+        import com.google.auth.oauth2.{
+          ImpersonatedCredentials,
+          ServiceAccountCredentials,
+          UserCredentials
+        }
+
+        _credentials match {
+          case sa: ServiceAccountCredentials => ("service account", sa.getAccount)
+          case uc: UserCredentials           => ("user", uc.getClientId)
+          case ic: ImpersonatedCredentials   => ("impersonated account", ic.getAccount)
+          case other: Credentials => ("user", s"unknown credential type ${other.getClass.getName}")
+        }
+      }
+
+      Try(fn(underlying).execute()) match {
+        case Success(response) => response
+        case Failure(e: GoogleJsonResponseException) if e.getStatusCode == 403 =>
+          val adc = getAuthenticatedUser
```

Comment:
[minor] this could be unpacked straight into (type, accountName) pair of variables:

```scala
val (accountType, accountName) = getAuthenticatedUser
```